[ST-NNNN] Polling Confirmations

younata opened the pull request:
New proposal for Swift Testing: Polling Confirmations.

These allow test authors to check that state updates over the course of some amount of time, in a very brute-force kind of way. Very helpful when there aren't other means of checking this available.

View [rendered proposal](https://github.com/younata/swift-evolution/blob/younata/testing-polling-expectations/proposals/testing/NNNN-polling-confirmations.md).

younata committed:
New pitch for testing: Polling Expectations

younata committed:
Add an alternatives considered section on having passesOnce continue to evaluate the expression after it passes

younata committed:
Add a link to the pitch thread

younata committed:
Set polling expectation's status to awaiting implementation

I'm working on it, but macros are hard.

younata committed:
Swift Testing Polling Expectations:

- Change the default timeout.
- Change how unexpected thrown errors are treated.
- Add future direction to add change monitoring via Observation
- Add alternative considered of Just Use A While Loop
- Add alternative considered for shorter default timeouts

younata committed:
Add a link to the PR containing polling expectations

younata committed:
Update polling proposal to refer to new confirmation functions

younata committed:
Polling Confirmations: Renamed from Polling Expectations

- Include commenting on why not pure counts (i.e. why counts + interval)
- Specify the configuration traits
- Specify callsite configuration of counts + interval

younata committed:
Polling Confirmations: Add a couple sentences on why pollingInterval must be positive

younata committed:
Polling Confirmations: Add requirePassesEventually and requireAlwaysPasses

younata committed:
Polling Confirmations: rewrite, again, for new API design

younata committed:
Mention the new Issue.Kind case, as well as why polling confirmations do not have a configurable clock

stmontgomery commented on proposals/testing/NNNN-polling-confirmations.md:
```diff
@@ -0,0 +1,512 @@
+/// complex scenarios where other forms of confirmation are insufficient. For
+/// example, waiting on some state to change that cannot be easily confirmed
+/// through other forms of `confirmation`.
+@available(_clockAPI, *)
```
`_clockAPI` is an availability definition specific to the Swift Testing codebase, and at build time it expands to actual OS version numbers. For the purpose of an API proposal I recommend showing the actual, expanded OS versions so people understand what they'll be in practice.

(This comment applies everywhere in the document which references `_clockAPI`.)

stmontgomery commented on proposals/testing/NNNN-polling-confirmations.md:
```diff
@@ -0,0 +1,512 @@
+
+### Platform Availability
+
+Polling confirmations will not be available on platforms that do not support
+Swift Concurrency.
```
The availability will also be constrained to OS versions which have the `Clock`, `Duration`, and related types. You may want to mention the specific Apple OS versions in particular.

stmontgomery commented on proposals/testing/NNNN-polling-confirmations.md:
```diff
@@ -0,0 +1,512 @@
+  /// The minimum amount of time to wait between polling attempts
+  public var interval: Duration?
+
+  public var isRecursive: Bool { true }
```
You can omit the `{ true }` _implementation_ and only show that it's a `get`-able property, e.g.

```suggestion
  public var isRecursive: Bool { get }
```

But I would recommend _documenting_ that this trait type is recursively applied to children. For other trait types we've done this in the DocC for the type itself.

stmontgomery commented on proposals/testing/NNNN-polling-confirmations.md:
```diff
@@ -0,0 +1,512 @@
+/// To add this trait to a test, use the
+/// ``Trait/pollingUntilFirstPassDefaults`` function.
+@available(_clockAPI, *)
+public struct PollingUntilFirstPassConfigurationTrait: TestTrait, SuiteTrait {
```
Assuming you do want to include the ability to influence this feature's behavior via a trait, I would suggest consolidating these two trait types into one if possible. For other traits we've tried to have the types be broader and use the `static` functions which initialize those types be the point where we differentiate their behavior. For example, the `.enabled(if:)` and `.disabled(if:)` functions both return a type of the same type: `ConditionTrait`.

But to accomplish this consolidation you'll likely need to store the `PollingStopCondition` in the trait instance. I think these changes would make for a simpler trait story for this feature: a single trait with a simpler name like `PollingConfigurationTrait`, with a property specifying its stopping condition.

stmontgomery commented on proposals/testing/NNNN-polling-confirmations.md:
````diff
@@ -0,0 +1,512 @@
+
+```swift
+public struct Issue {
+  public enum Kind {
````
Small style suggestion: When proposing a new member of an existing type, such as a new case in an enum here, I often add a placeholder comment like `// ...` above the additions to reflect that there are existing members and you're just _adding_ things.

stmontgomery commented on proposals/testing/NNNN-polling-confirmations.md:
````diff
@@ -0,0 +1,512 @@
+
+```swift
+/// A type describing an error thrown when polling fails.
+public struct PollingFailedError: Error, Sendable, CustomIssueRepresentable {}
````
`CustomIssueRepresentable` is not a `public` protocol so it doesn't need to be shown here and I'd omit it to avoid confusion.

stmontgomery commented on proposals/testing/NNNN-polling-confirmations.md:
```diff
@@ -0,0 +1,512 @@
+@available(_clockAPI, *)
+public struct PollingUntilStopsPassingConfigurationTrait: TestTrait, SuiteTrait {
+  /// How long to continue polling for
+  public var duration: Duration?
+  /// The minimum amount of time to wait between polling attempts
+  public var interval: Duration?
```
Since both of these properties are optional, can you document (and decide, if it hasn't been settled yet) what the behavior will be if either or both of these are `nil`?

stmontgomery commented on proposals/testing/NNNN-polling-confirmations.md:
```diff
@@ -0,0 +1,512 @@
+
+@available(_clockAPI, *)
+extension Trait where Self == PollingUntilFirstPassConfigurationTrait {
+  /// Specifies defaults for ``confirmPassesEventually`` in the test or suite.
```
This documentation still references the (older, I think?) API name `confirmPassesEventually`. (Similar comment for the other trait below, referencing `confirmPassesAlways`.)

stmontgomery commented on proposals/testing/NNNN-polling-confirmations.md:
```diff
@@ -0,0 +1,512 @@
+### Default Polling Configuration
+
+For all polling confirmations, the Testing library will default `duration` to
+1 second, and `interval` to 1 millisecond.
```
I'm somewhat hesitant to formally document and "promise" particular values for the duration and polling interval. Moreso for the latter; those are effectively policy details and I could imagine in the future having those be platform- or environment-specific. Or we could get fancier and decide to make the polling interval a curve rather than a fixed value: for example it could poll very frequently and aggressively at the beginning, but over time gradually back off if it seems like the condition is taking a while to become satisfied.

These are just speculative ideas and I'm not suggesting you actually pursue them right now. (Maybe they'd be good to mention in Alternatives considered?) But for now, I'm only suggesting that if we formally specify and commit to those numbers in this proposal, we might lose the flexibility to refine the behavior later. So maybe it would be better to be a little more vague, and say that the default interval will be short, but the testing library may adjust it to behave well in each runtime environment.

stmontgomery commented on proposals/testing/NNNN-polling-confirmations.md:
````diff
@@ -0,0 +1,512 @@
+}
+```
+
+With the definition of `Aquarium` above, the closure will only need to be
````
nit: "may" only need? Since technically you can't predict or guarantee any particular timing; the system could be under heavy load
```suggestion
With the definition of `Aquarium` above, the closure may only need to be
```

stmontgomery commented on proposals/testing/NNNN-polling-confirmations.md:
```diff
@@ -0,0 +1,512 @@
+evaluated a few times before it starts returning true. At which point polling
+will end, and no failure will be reported.
+
+Polling will be stopped in the following cases:
+
+- The specified `duration` has elapsed.
+- If the task that started the polling is cancelled.
+- For `PollingStopCondition.firstPass`: The first time the closure returns true
+  or a non-nil value
+- For `PollingStopCondition.stopsPassing`: The first time the closure returns
+  false or nil.
+- The first time the closure throws an error.
```
It might not be clear to a reader that this is an "OR" list not "AND". I find it helpful when proposals clarify that kind of thing by adding _"…, or"_ to the second-to-last bullet. For example:

```suggestion
Polling will be stopped when either:

- the specified duration has elapsed,
- the task that started the polling is cancelled,
- the closure returns a value that satisfies the stopping condition, or
- the closure throws an error.
```

(And in that suggestion I shortened the third bullet, instead of reiterating the stopping condition behaviors, since those are specified elsewhere in the document.)

younata committed:
PR Feedback

younata committed:
Polling confirmations: expand the motivation section, adding a section on when polling should not be used.

younata committed:
Polling confirmations: Expand on why directly using duration doesn't work.

younata committed:
Update proposals/testing/NNNN-polling-confirmations.md

Co-authored-by: Jonathan Grynspan <grynspan@me.com>

younata committed:
Testing/Polling Confirmations: Add a failure reason API

younata committed:
PollingFailureReason is now nested inside of PollingFailedError

younata committed:
Update documentation comments in the code samples

younata committed:
Testing: Link the second pitch thread for polling confirmations

younata committed:
Update documentation comment for PollingConfirmationConfigurationTrait to mention that duration may be greater than or equal to interval

younata commented:
> Looks like there's some merge or rebase issue, files besides the new proposal are being shown as modified

Whoops. That's corrected now. I wonder how I messed up that rebase.

younata committed:
Add a Future Directions section, discussing tracking task execution time

Minor grammar improvements

ktoso commented on proposals/testing/NNNN-polling-confirmations.md:
```diff
@@ -0,0 +1,668 @@
+  until stopCondition: PollingStopCondition,
+  within duration: Duration? = nil,
+  pollingEvery interval: Duration? = nil,
+  isolation: isolated (any Actor)? = #isolation,
```
Prefer expressing this by making the function “nonisolated(nonsending)”

younata replied:
Thanks! I clearly missed the SE proposal that changed this.

ktoso commented on proposals/testing/NNNN-polling-confirmations.md:
```diff
@@ -0,0 +1,668 @@
+  pollingEvery interval: Duration? = nil,
+  isolation: isolated (any Actor)? = #isolation,
+  sourceLocation: SourceLocation = #_sourceLocation,
+  _ body: @escaping () async throws -> Bool
```
Should this be nonisolated nonsending as well?

younata committed:
Add more detail on why task execution time would improve polling

Also add a couple paragraphs discussing known delays in code when using polling.

younata committed:
Annotation polling confirmation bodies as nonisolated(nonsending)

Keep the @escaping, even though that's not strictly necessary: it's less work to specify escaping now and later remove it
than it is to introduce the escaping restriction later.

younata committed:
Testing/Polling Confirmations: Move tracking test execution time to an alternatives considered.

As it turns out, this approach isn't feasible. Sadly.

maartene committed:
Update NNNN-polling-confirmations.md

Corrected a couple of typos

younata committed:
More spelling corrections

younata committed:
ST Polling Confirmations: Describe mechanism to relay information about why a polling attempt failed

younata committed:
Include a concrete example of recording to a log file

stmontgomery commented on proposals/testing/NNNN-polling-confirmations.md:
```diff
@@ -0,0 +1,1051 @@
+  pollingEvery interval: Duration? = nil,
+  sourceLocation: SourceLocation = #_sourceLocation,
+  _ body: nonisolated(nonsending) @escaping () async throws -> PollResult<Bool>
+) async throws {
+  let poller = Poller(
+    stopCondition: stopCondition,
+    duration: stopCondition.duration(with: duration),
+    interval: stopCondition.interval(with: interval),
+    comment: comment,
+    sourceContext: SourceContext(
+      backtrace: .current(),
+      sourceLocation: sourceLocation
+    )
+  )
+  try await poller.evaluate() {
+    do {
+      return try await body()
+    } catch {
+      return false
+    }
+  }
+}
```
The implementation of the function can be omitted from the proposal

stmontgomery commented on proposals/testing/NNNN-polling-confirmations.md:
```diff
@@ -0,0 +1,1051 @@
+`confirmation(_:until:within:pollingEvery:sourceLocation:_:)`. These
+functions take in a closure to be repeatedly evaluated until the specific
+condition passes, waiting at least some amount of time - specified by
+`pollingEvery`/`interval` and defaulting to 1 millisecond - before evaluating
```
I think it would be better to not explicitly specify what the default polling interval will be, because it may need to vary per-platform or be dynamic (which is related to the idea I suggested a while back which led to the "Curved polling rates" future direction discussed below).

In particular, I think 1 millisecond may be too aggressive in practice, especially when parallelization is enabled. By comparison, XCTest (in its [`XCTWaiter` implementation](https://github.com/swiftlang/swift-corelibs-xctest/blob/cc562f728ece742925b75216ee5a16748655fea8/Sources/XCTest/Public/Asynchronous/XCTWaiter.swift#L425)) uses a polling interval of 0.1 seconds, i.e. 100ms, and it never even has more than one test running at a time in the same process.

So while I think the implementation of this feature likely should poll less frequently than 1ms, when it comes to this _proposal_ my feedback is that a default value should simply be left undefined. The proposal could say something less committal instead, like,

> The specific default polling interval is platform- and environment-dependent, and may change dynamically in response to system load. But its default value will typically be 1 second or less.

stmontgomery commented on proposals/testing/NNNN-polling-confirmations.md:
```diff
@@ -0,0 +1,1051 @@
+  pollingEvery interval: Duration? = nil,
+  sourceLocation: SourceLocation = #_sourceLocation,
+  _ body: nonisolated(nonsending) @escaping () async throws -> sending PollResult<R>
+) async throws -> R {
+  let poller = Poller(
+    stopCondition: stopCondition,
+    duration: stopCondition.duration(with: duration),
+    interval: stopCondition.interval(with: interval),
+    comment: comment,
+    sourceContext: SourceContext(
+      backtrace: .current(),
+      sourceLocation: sourceLocation
+    )
+  )
+  return try await poller.evaluateOptional() {
+    do {
+      return try await body()
+    } catch {
+      return nil
+    }
+  }
+}
```
Function body can be omitted here as well
